### PR TITLE
Improve over(under)lay brushes

### LINKED ISSUE #280: Blocks float in the air when doing /b overlay all
### Server Implementation

Paper

### Server Version

1.20

### Describe the bug

I found out why it is that weird with the fern in the air, it is because of "all" in the command. If I do "some", all blocks are on top as they should.

### To Reproduce

1. take a gunpowder and use /b overlay some (the depth is 1)
2. Type in chat /v fern (for example)
3. Use the brush on natural blocks, it looks fine
4. Change the brush to /b overlay all
5. Use the brush on not natural blocks, they float on the corner of a higher block

### Expected behaviour

That it behaves like the /b overlay some but on not natural blocks (air should be excluded as a possible block so it might not float)

### Screenshots / Videos

grafik

### Error log (if applicable)

There are no errors, unfortunately

### Favs Debugpaste

https://athion.net/ISPaster/paste/view/fd8ffe5669f44bc7bba132e77bc2e490

### Favs Version

2.9.3

### Checklist

- [X] I have included a Fawe debugpaste.
- [X] I am using the newest build from https://dev.bukkit.org/projects/favs and the issue still persists.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Aurelien30000**
## Overview

Fixes #280

## Description

- Use boolean array instead of int array to keep track of already found blocks.
- Add origin air checks (as suggested by @OneLiteFeather) for overlay brushes when all blocks parameter is being used.

```[tasklist]
### Submitter Checklist
- [x] Make sure you are opening from a topic branch (**/feature/fix/docs/ branch** (right side)) and not your main branch.
- [x] Ensure that the pull request title represents the desired changelog entry.
- [x] New public fields and methods are annotated with `@since TODO`.
- [x] I read and followed the [contribution guidelines](https://github.com/IntellectualSites/.github/blob/main/CONTRIBUTING.md).
```

### CODE REVIEW

File: src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/OverlayBrush.java [hunk 1]
```diff
@@ -156,7 +156,7 @@ private void overlayTwo(Snipe snipe) {
                 int blockY = targetBlock.getY();
                 int blockZ = targetBlock.getZ();
                 for (int y = blockY; y >= editSession.getMinY() && !surfaceFound; y--) { // start scanning from the height you clicked at
-                    if (memory[x + brushSize][z + brushSize] != 1) { // if haven't already found the surface in this column
+                    if (!memory[x + brushSize][z + brushSize]) { // if haven't already found the surface in this column
```
Comment (TheMeinerLP) [inline src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/OverlayBrush.java]: The `!` can be overread, maybe a variable helps here :D 
e.x:
```java
var noSurfaceFound = !memory[x + brushSize][z + brushSize];
if(noSurfaceFound) // ...
```

File: src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/OverlayBrush.java [hunk 2]
```diff
@@ -169,21 +169,24 @@ private void overlayTwo(Snipe snipe) {
...
-                                            this.performer.perform(
-                                                    getEditSession(),
-                                                    blockX + x,
-                                                    clampY(y + index),
-                                                    blockZ + z,
-                                                    this.clampY(blockX + x, y + index, blockZ + z)
-                                            ); // fills down as many layers as you specify in
-                                            // parameters
-                                            memory[x + brushSize][z + brushSize] = 1; // stop it from checking any other blocks in this vertical 1x1 column.
+                                        if (!Materials.isEmpty(type)) {
```
Comment (TheMeinerLP) [inline src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/OverlayBrush.java]: see above comment

File: src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/UnderlayBrush.java [hunk 3]
```diff
@@ -98,35 +100,37 @@ private void underlay(Snipe snipe) {
                 int blockY = targetBlock.getY();
                 int blockZ = targetBlock.getZ();
                 for (int y = blockY; y < blockY + this.depth; y++) { // start scanning from the height you clicked at
-                    if (memory[x + brushSize][z + brushSize] != 1) { // if haven't already found the surface in this column
+                    if (!memory[x + brushSize][z + brushSize]) { // if haven't already found the surface in this column
```
Comment (TheMeinerLP) [inline src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/UnderlayBrush.java]: see above comment

File: src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/UnderlayBrush.java [hunk 4]
```diff
@@ -150,33 +154,38 @@ private void underlay2(Snipe snipe) {
...
```
Comment (TheMeinerLP) [inline src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/UnderlayBrush.java]: see above comment

File: src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/UnderlayBrush.java [hunk 5]
```diff
@@ -150,33 +154,38 @@ private void underlay2(Snipe snipe) {
...
-                                    this.performer.perform(
-                                            getEditSession(),
-                                            blockX + x,
-                                            clampY(y - i),
-                                            blockZ + z,
-                                            clampY(blockX + x, y - i, blockZ + z)
-                                    ); // fills down as many layers as you specify in
-                                    // parameters
-                                    memory[x + brushSize][z + brushSize] = 1; // stop it from checking any other blocks in this vertical 1x1 column.
+                                if (!Materials.isEmpty(type)) {
```
Comment (TheMeinerLP) [inline src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/UnderlayBrush.java]: see above comment

File: src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterOverlayBrush.java [hunk 6]
```diff
@@ -209,7 +209,7 @@ private void splatterOverlay(Snipe snipe) {
                 int blockZ = targetBlock.getZ();
                 for (int y = targetBlock.getY(); y >= editSession.getMinY(); y--) {
                     // start scanning from the height you clicked at
-                    if (memory[x + brushSize][z + brushSize] != 1) {
+                    if (!memory[x + brushSize][z + brushSize]) {
```
Comment (TheMeinerLP) [inline src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterOverlayBrush.java]: see above comment

File: src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterOverlayBrush.java [hunk 7]
```diff
@@ -311,19 +311,17 @@ private void splatterOverlayTwo(Snipe snipe) {
...
                 BlockVector3 targetBlock = this.getTargetBlock();
                 int blockX = targetBlock.getX();
                 int blockZ = targetBlock.getZ();
                 for (int y = targetBlock.getY(); y >= editSession.getMinY(); y--) { // start scanning from the height you clicked at
-                    if (memory[x + brushSize][z + brushSize] != 1) { // if haven't already found the surface in this column
-                        if ((Math.pow(x, 2) + Math.pow(
-                                z,
-                                2
-                        )) <= brushSizeSquared && splat[x + brushSize][z + brushSize] == 1) { // if inside of the column...&& if to be splattered
+                    if (!memory[x + brushSize][z + brushSize]) { // if haven't already found the surface in this column
```
Comment (TheMeinerLP) [inline src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterOverlayBrush.java]: see above comment

File: src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterOverlayBrush.java [hunk 8]
```diff
@@ -335,36 +333,37 @@ private void splatterOverlayTwo(Snipe snipe) {
...
-                                            this.performer.perform(
-                                                    getEditSession(),
-                                                    blockX + x,
-                                                    clampY(y + i + this.yOffset),
-                                                    blockZ + z,
-                                                    clampY(blockX + x, y + i + this.yOffset, blockZ + z)
-                                            ); // fills down as many layers as you specify in
-                                            // parameters
-                                            memory[x + brushSize][z + brushSize] = 1; // stop it from checking any other blocks in this vertical 1x1 column.
+                                        if (!Materials.isEmpty(type)) {
```
Comment (TheMeinerLP) [inline src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/splatter/SplatterOverlayBrush.java]: see above comment

**Review by TheMeinerLP — CHANGES_REQUESTED** [review]
This improves the readability for external/other developers.

File: src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/OverlayBrush.java (hunk 1 above)
Comment (Aurelien30000) [reply src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/OverlayBrush.java]: FAVS is full of overlapping code horrors :sweat_smile: But yeah, let's improve it step by step, thanks for the review.

**Review by TheMeinerLP — APPROVED** [review]
LGTM

**Review by NotMyFault — APPROVED** [review]

**Pull request merged**